### ISSUE TITLE
Tutorial notebook(s) for truth catalog

### ISSUE THREAD

**Issue by yymao** [LSSTDESC/DC2_Repo#184]
Add tutorial notebook(s) for the truth catalog. The tutorial should use the reader, which does not yet exist (see https://github.com/LSSTDESC/gcr-catalogs/issues/130). 

We should also replace the reference catalog used in @slosar and students' matching notebook with the truth catalog (with GCR this should be just changing one line).

**Comment by danielsf** [comment]
I'm almost ready to incorporate the truth catalog for static sources into the GCR framework.  These are the columns I imagine being included in the truth catalog.  Please pipe-up if I am neglecting something you think is important (@slosar @wmwv @cwwalter )

ra, dec -- in degrees

healpix_id  -- the index of the n_side=2048 nested healpixel containing the source

redshift

is_star -- an int (either 1 or 0)

has_agn -- an int (either 1 or 0)

is_sprinkled -- an int (either 1 or 0); objects with is_sprinkled==1 will have no counterpart in the extragalactic catalog

ugrizy -- the magnitudes of the source in all 6 LSST filters.  These magnitudes will be the mean magnitudes for variable sources.  For galaxies, they will include all components (bulge, disk, AGN).  No dust extinction will be applied, since Run 1.1p was run with a version of PhoSim that had accidentally turned off all dust.

**Comment by danielsf** [comment]
also: galaxy_id, which I will set to -1 for stars

**Comment by danielsf** [comment]
While our Milky Way model does have radial velocity, that information never makes it to ImSim/PhoSim, so the stars are at `redshift==0`

The sprinkler does assign a new `galaxy_id` for the purposes of generating the object identifier that gets passed to PhoSim, so `galaxy_id>0` for sprinkled objects.

Also: I remembered that the stars also have a unique identifier (`simobjid`) within the CatSim model, so that column will be `object_id`, which will be `==galaxy_id` in the case of galaxies and `==simobjid` in the case of stars.  Users will be have to use the `star` and `sprinkled` columns to figure out what ID they are actually looking at.

**Comment by danielsf** [comment]
PR #198 opened for anyone interested

**Comment by yymao** [comment]
Now that we have truth catalogs, should we update the matching notebook(s) to use truth instead of reference? What do you think, @danielsf @wmwv @slosar?

**Comment by wmwv** [comment]
Yes

**Comment by danielsf** [comment]
Preemptive note to whoever takes on that work: the truth catalog does not work with `GCRQuery` (`GCRQuery` seems designed for querying pandas dataframes or astropy tables; the truth catalog is, ultimately, a sqlite database).  When I was testing the truth catalog with HyeYun and Anze's notebooks, I had to replace the reliance on `GCRQuery` with the use of the `native_filters` kwarg.

**Comment by yymao** [comment]
@danielsf hmm, but `GCRQuery` should work on the returned object, which should be just a python dict of numpy arrays. Unlike native filters, `filters` do not need to know the underlying format. 

Does the reader not return such object? If so, we should change the reader.

**Comment by danielsf** [comment]
The reader should return a dict of numpy arrays.  When I tried to pass it a `GCRQuery` as a filter, I got an error.  I do not remember which error.  When I looked at the underlying `easyquery` structure, it looked to me like it was expecting to be handling dataframes and astropy tables, but I was not 100% sure what I was looking at and didn't put much time into it.

**Comment by yymao** [comment]
Yes, `easyquery` works for numpy recarray, astropy table, and pandas dataframe, but `GCRQuery` is a subclass of `easyquery`and it's designed to work on all GCR returns. If it is not working then it's most likely something wrong with the reader. 

@danielsf any chance you can reproduce the error, or recall which notebook you were trying?

**Comment by danielsf** [comment]
I was hacking this notebook

https://github.com/LSSTDESC/DC2_Repo/blob/master/Notebooks/FoFCatalogMatching%20Histogram.ipynb

to use the new truth catalog, rather than the reference catalog.

**Comment by danielsf** [comment]
@yymao At what point is the reader supposed to return a dict?  Is the dict supposed to be what comes out of `_iter_native_quantities`?

**Comment by cwwalter** [comment]
Yes on using truth.  What about the case where the flux is changed by @EiffL's knots?  I think this will be the 1st case of others more complicated morphologies so we might want to consider the best way to handle this generally.

**Comment by danielsf** [comment]
The knots might actually be easy since they are implemented in a flux-conserving way, so there is no difference in total flux between a galaxy simulated as bulge+disk and a galaxy simulated as bulge+disk+knots.

There are, however, more design decisions to be made.  The current truth catalog does nothing with time dependence.  This is literally just the simplest possible truth catalog that could be banged out in a week.

**Comment by cwwalter** [comment]
Oh, I missed that the individual components won't be listed. Yes, for right now for the knots there should be no issue though.

One thing, I forgot to mention above: For imSim we will be properly applying extinction.  So, I think we want both extincted and unextincted magnitudes so the truth catalog can be used with both simulations. That's probably interesting anyway.

**Comment by yymao** [comment]
Now that the truth catalog filtering issue is fixed (see https://github.com/LSSTDESC/gcr-catalogs/pull/153 and https://github.com/LSSTDESC/sims_GCRCatSimInterface/pull/54), I'll update the `FoFCatalogMatching Histogram.ipynb` notebook under `u/yymao/use-truth-for-matching`.
